File: 0113-path-sum-ii/0113-path-sum-ii.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private List<List<Integer>> result;

    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        result = new ArrayList<>();

        if (root == null) return result;

        dfs(root, new ArrayList<>(), 0, targetSum);

        return result;
    }

    private void dfs(TreeNode node, List<Integer> curPath, int curSum, int target) {
        curPath.add(node.val);
        curSum += node.val;

        boolean isLeaf = node.left == null && node.right == null;
        if (isLeaf && curSum == target) {
            result.add(new ArrayList<>(curPath));
        }

        if (node.left != null) {
            dfs(node.left, curPath, curSum, target);
        }

        if (node.right != null) {
            dfs(node.right, curPath, curSum, target);
        }

        curPath.remove(curPath.size() - 1);
    }
}

File: 0113-path-sum-ii/0113-path-sum-ii.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    private val result = mutableListOf<List<Int>>()

    fun pathSum(root: TreeNode?, targetSum: Int): List<List<Int>> {
        root?.let { dfs(it, mutableListOf<Int>(), 0, targetSum) }
        return result
    }

    private fun dfs(node: TreeNode, curPath: MutableList<Int>, curSum: Int, target: Int) {
        curPath.add(node.`val`)
        val newSum = curSum + node.`val`

        if (node.left == null && node.right == null && newSum == target) {
            result.add(curPath.toList())
        }

        node.left?.let { dfs(it, curPath, newSum, target) }
        node.right?.let { dfs(it, curPath, newSum, target) }

        curPath.removeLast()
    }
}